### PR TITLE
Migrating unit tests to pytest (Part 5)

### PR DESCRIPTION
**Pull request by prady0t**

### CODE REVIEW

File: tests/unit/test_discretisations/test_discretisation.py [hunk 1]
```diff
@@ -817,7 +811,7 @@ def test_broadcast(self):
             7 * np.ones_like(mesh["negative electrode"].edges[:, np.newaxis]),
         )
 
-    def test_broadcast_2D(self):
+    def test_broadcast_2_d(self):
```
Comment (Saransh-cpp) [inline tests/unit/test_discretisations/test_discretisation.py:814]: ?
```suggestion
    def test_broadcast_2d(self):
```

File: tests/unit/test_discretisations/test_discretisation.py [hunk 2]
```diff
@@ -842,20 +836,20 @@ def test_broadcast_2D(self):
...
+            1,
         )
         y_test = np.linspace(0, 1, mesh["current collector"].npts)
         np.testing.assert_array_equal(
             broad_to_edges_disc.evaluate(y=y_test),
             np.outer(y_test, np.ones(mesh["separator"].npts + 1)).reshape(-1, 1),
         )
 
-    def test_secondary_broadcast_2D(self):
+    def test_secondary_broadcast_2_d(self):
```
Comment (Saransh-cpp) [inline tests/unit/test_discretisations/test_discretisation.py:852]:
```suggestion
    def test_secondary_broadcast_2d(self):
```

File: tests/unit/test_discretisations/test_discretisation.py [hunk 3]
```diff
@@ -864,28 +858,28 @@ def test_secondary_broadcast_2D(self):
...
+        assert isinstance(broad_to_edges_disc, pybamm.MatrixMultiplication)
+        assert isinstance(broad_to_edges_disc.children[0], pybamm.Matrix)
+        assert isinstance(broad_to_edges_disc.children[1], pybamm.StateVector)
+        assert broad_to_edges_disc.shape == (
+            mesh["negative particle"].npts * (mesh["negative electrode"].npts + 1),
+            1,
         )
 
-    def test_tertiary_broadcast_3D(self):
+    def test_tertiary_broadcast_3_d(self):
```
Comment (Saransh-cpp) [inline tests/unit/test_discretisations/test_discretisation.py:882]:
```suggestion
    def test_tertiary_broadcast_3d(self):
```

File: tests/unit/test_discretisations/test_discretisation.py [hunk 4]
```diff
@@ -960,7 +948,7 @@ def test_concatenation_of_scalars(self):
         )[:, np.newaxis]
         np.testing.assert_allclose(eqn_disc.evaluate(), expected_vector)
 
-    def test_concatenation_2D(self):
+    def test_concatenation_2_d(self):
```
Comment (Saransh-cpp) [inline tests/unit/test_discretisations/test_discretisation.py:951]:
```suggestion
    def test_concatenation_2d(self):
```

File: tests/unit/test_batch_study.py [hunk 5]
```diff
@@ -41,62 +41,62 @@ def test_solve(self):
...
-            with self.assertRaises(ValueError):
+            with pytest.raises(ValueError):
                 pybamm.BatchStudy(
                     models={"SPM": spm, "SPM uniform": spm_uniform}, **{name: {None}}
                 )
 
         # Tests for None when only models are given with permutations=False
         bs_false_only_models.solve(t_eval=[0, 3600])
-        self.assertEqual(2, len(bs_false_only_models.sims))
+        assert 2 == len(bs_false_only_models.sims)
```
Comment (Saransh-cpp) [inline tests/unit/test_batch_study.py:51]:
```suggestion
        assert  len(bs_false_only_models.sims) == 2
```

File: tests/unit/test_batch_study.py [hunk 6]
```diff
@@ -41,62 +41,62 @@ def test_solve(self):
...
 
         # Tests for None when only models are given with permutations=True
         bs_true_only_models.solve(t_eval=[0, 3600])
-        self.assertEqual(2, len(bs_true_only_models.sims))
+        assert 2 == len(bs_true_only_models.sims)
```
Comment (Saransh-cpp) [inline tests/unit/test_batch_study.py:55]: ditto

File: tests/unit/test_batch_study.py [hunk 7]
```diff
@@ -41,62 +41,62 @@ def test_solve(self):
...
 
         # Tests for BatchStudy when permutations=False
         bs_false.solve()
         bs_false.plot(show_plot=False)
-        self.assertEqual(2, len(bs_false.sims))
+        assert 2 == len(bs_false.sims)
```
Comment (Saransh-cpp) [inline tests/unit/test_batch_study.py:60]: ditto

File: tests/unit/test_batch_study.py [hunk 8]
```diff
@@ -41,62 +41,62 @@ def test_solve(self):
...
                 experiment.steps for experiment in bs_false.experiments.values()
             ]
-            self.assertIn(output_experiment, experiments_list)
+            assert output_experiment in experiments_list
 
         # Tests for BatchStudy when permutations=True
         bs_true.solve()
         bs_true.plot(show_plot=False)
-        self.assertEqual(4, len(bs_true.sims))
+        assert 4 == len(bs_true.sims)
```
Comment (Saransh-cpp) [inline tests/unit/test_batch_study.py:79]: ditto

File: tests/unit/test_expression_tree/test_operations/test_evaluate_python.py [hunk 9]
```diff
@@ -229,17 +227,17 @@ def test_domain_concatenation(self):
...
+            list(variable_symbols.values())[2]
+            == f"np.concatenate(({a0_str},{b0_str},{b1_str}))"
         )
 
         evaluator = pybamm.EvaluatorPython(expr)
         result = evaluator(y=y)
         np.testing.assert_allclose(result, expr.evaluate(y=y))
 
-    def test_domain_concatenation_2D(self):
+    def test_domain_concatenation_2_d(self):
```
Comment (Saransh-cpp) [inline tests/unit/test_expression_tree/test_operations/test_evaluate_python.py:240]:
```suggestion
    def test_domain_concatenation_2d(self):
```

**Review by Saransh-cpp — APPROVED** [review]
Thank, @prady0t! The comments are just personal opinions / usual practices.

File: tests/unit/test_discretisations/test_discretisation.py (hunk 1 above)
Comment (agriyakhetarpal) [reply tests/unit/test_discretisations/test_discretisation.py:814]: I think this was discussed in an earlier PR, it's a `pytestify` bug 🤔

File: tests/unit/test_citations.py
Comment (agriyakhetarpal) [inline tests/unit/test_citations.py:1]: This is an example of a file that contains a lot of boilerplate and can be reduced by fixtures and parametrizing the test cases. You can add files like these under a separate heading in the tracking issue.

File: tests/unit/test_discretisations/test_discretisation.py (hunk 1 above)
Comment (Saransh-cpp) [reply tests/unit/test_discretisations/test_discretisation.py:814]: ahh, then please feel free to skip

**Review by agriyakhetarpal — APPROVED** [review]
Thanks, @prady0t! This can be merged once @Saransh-cpp's comments are resolved.

File: tests/unit/test_citations.py
Comment (prady0t) [reply tests/unit/test_citations.py:1]: Done!

**Pull request merged**